src/MagnumPlugins/DdsImporter/DdsImporter.cpp

```diff
@@ -0,0 +1,351 @@
+               ddsh.ddspf.bBitMask == 0x00FF0000 &&
+               ddsh.ddspf.aBitMask == 0xFF000000) {
+        _colorFormat.uncompressed = ColorFormat::RGBA;
+    } else if (ddsh.ddspf.rgbBitCount == 24 &&
+               ddsh.ddspf.rBitMask == 0x000000FF &&
+               ddsh.ddspf.gBitMask == 0x0000FF00 &&
+               ddsh.ddspf.bBitMask == 0x00FF0000) {
+        _colorFormat.uncompressed = ColorFormat::RGB;
+        _components = 3;
+    } else if (ddsh.ddspf.rgbBitCount == 24 &&
+               ddsh.ddspf.rBitMask == 0x00FF0000 &&
+               ddsh.ddspf.gBitMask == 0x0000FF00 &&
+               ddsh.ddspf.bBitMask == 0x000000FF) {
+        _colorFormat.uncompressed = ColorFormat::BGR;
+        _components = 3;
+    } else if (ddsh.ddspf.rgbBitCount == 8) {
+        #ifndef MAGNUM_TARGET_GLES2
+        _colorFormat.uncompressed = ColorFormat::Red;
+        #else
+        _colorFormat.uncompressed = ColorFormat::Luminance;
+        #endif
+    } else {
+        Error() << "Unknow texture format";
+    }
+
+    const Vector3i size{Int(ddsh.width), Int(ddsh.height), Int(Math::max(ddsh.depth, 1u))};
+
+    /* check how many mipmaps to load */
+    const UnsignedInt numMipmaps = (ddsh.flags & DdsDescriptionFlag::MipMapCount)
+                    ? ((ddsh.mipMapCount == 0) ? 0 : ddsh.mipMapCount - 1)
+                    : 0;
+
+    /* load all surfaces for the image (6 surfaces for cubemaps) */
+    const UnsignedInt numImages = (isCubemap) ? 6 : 1;
+    size_t offset = 128;
+    for (UnsignedInt n = 0; n < numImages; ++n) {
+        offset = addImageDataOffset(size, offset);
+
+        Vector3i mipSize{size};
+
+        /* load all mipmaps for current surface */
+        for (UnsignedInt i = 0; i < numMipmaps && (mipSize.x() || mipSize.y()); i++) {
+            /* shrink to next power of 2 */
+            mipSize = Math::max(mipSize >> 1, Vector3i{1});
+
+            offset = addImageDataOffset(mipSize, offset);
+        }
+    }
+
+}
+
+size_t DdsImporter::addImageDataOffset(const Vector3i& dims, const size_t offset)
+{
+    if (_compressed) {
+        const unsigned int size = (dims.z()*((dims.x() + 3)/4)*(((dims.y() + 3)/4))*((_colorFormat.compressed == CompressedColorFormat::RGBAS3tcDxt1) ? 8 : 16));
+
+        _imageData.push_back({dims, _in.slice(offset, offset + size)});
+
+        return offset + size;
+    } else {
+        const unsigned int numPixels = dims.product();
+        const unsigned int size = numPixels * _components;
+
+        _imageData.push_back({dims, _in.slice(offset, offset + size)});
+
+        return offset + size;
+    }
+}
+
+UnsignedInt DdsImporter::doImage2DCount() const {  return (_volume) ? 0 : _imageData.size(); }
+
+std::optional<ImageData2D> DdsImporter::doImage2D(UnsignedInt id) {
+    ImageDataOffset& dataOffset = _imageData[id];
+
+    // copy image data
+    Containers::Array<char>* data = new Containers::Array<char>(dataOffset._data.size());
```

Comment:
ImageData2D doesn't data, right? Don't I need to make sure it stays around? For the uncompressed case, I'm moving data, which would, of course make the `new` superfluous. There does not seem to be a constructor to move data into ImageData2D when uncompressed.